packages/preview2-shim/lib/nodejs/filesystem.js

```diff
@@ -1,78 +1,424 @@
-import { openSync, constants, statSync, lstatSync, fstatSync, closeSync, readdirSync } from 'node:fs';
-import { _createFsStream, _dropFsStream, _getFsStreamContext } from './io.js';
-
-// default is full permissions
-let preopenCnt = 4;
-export let _descriptors = {
-  3: { type: 'directory', path: '/', parent: null, subpathTypes: {} }
-};
-let directories = [[3, '/']];
-
-export function _setPreopens (preopens) {
-  _descriptors = {};
-  directories = [,,];
-  for (const [virtualPath, path] of Object.entries(preopens)) {
-    _descriptors[preopenCnt] = { type: 'directory', path, parent: null, subpathTypes: {} };
-    directories.push([preopenCnt++, virtualPath]);
+import { _io } from './io.js';
+import { environment } from './cli.js';
+import { constants, readSync, openSync, opendirSync, closeSync, fstatSync, lstatSync, statSync, writeSync } from 'node:fs';
+import { platform } from 'node:process';
+
+const isWindows = platform === 'win32';
+
+class ReadableFileStream {
+  constructor (hostFd, position) {
+    this.hostFd = hostFd;
+    this.position = position;
+  }
+  read (len) {
+    const buf = new Uint8Array(Number(len));
+    const bytesRead = readSync(this.hostFd, buf, this.position ? { position: this.position } : undefined);
```

Comment:
We should not be using the internal position here, but track the position as an offset within the stream itself.